source/non_matching/fe_immersed_values.cc

```diff
@@ -0,0 +1,216 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 - 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/non_matching/fe_immersed_values.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+namespace NonMatching
+{
+  template <int dim>
+  FEImmersedSurfaceValues<dim>::FEImmersedSurfaceValues(
+    const Mapping<dim> &                  mapping,
+    const FiniteElement<dim> &            element,
+    const ImmersedSurfaceQuadrature<dim> &quadrature,
+    const UpdateFlags                     update_flags)
+    : FEValuesBase<dim, dim>(quadrature.size(),
+                             element.dofs_per_cell,
+                             update_default,
+                             mapping,
+                             element)
+    , quadrature(quadrature)
+  {
+    initialize(update_flags);
+  }
+
+
+
+  template <int dim>
+  void
+  FEImmersedSurfaceValues<dim>::reinit(
+    const typename Triangulation<dim>::cell_iterator &cell)
+  {
+    // Check that mapping and reference cell type are compatible:
+    Assert(this->get_mapping().is_compatible_with(cell->reference_cell()),
+           ExcMessage(
+             "You are trying to call FEValues::reinit() with a cell of type " +
+             cell->reference_cell().to_string() +
+             " with a Mapping that is not compatible with it."));
+
+    // No FE in this cell, so no assertion necessary here.
+    this->maybe_invalidate_previous_present_cell(cell);
```

Comment:
I am wondering if we should be explicit here and prevent users to reuse this class:
```suggestion
    Assert(this->present_cell==false, ExcMessage("This class should not be reused."));
```

Reply:
Yes. It's probably a mistake if you call `reinit` twice.